src/compiler/semantic/rules.rs

```diff
@@ -1062,4 +1138,77 @@ mod test {
             r#"[SemErr { msg: "Cannot use true in expression 2 + true", dsym: DebugSymRef { line: 15, cols: "42-46" } }, SemErr { msg: "Cannot use true in expression 1 * true", dsym: DebugSymRef { line: 32, cols: "24-28" } }, SemErr { msg: "Cannot use false in expression false - 123", dsym: DebugSymRef { line: 32, cols: "31-36" } }, SemErr { msg: "Cannot use false in expression false * false", dsym: DebugSymRef { line: 32, cols: "50-55" } }, SemErr { msg: "Cannot use false in expression false * false", dsym: DebugSymRef { line: 32, cols: "58-63" } }]"#
         );
     }
+
+    #[test]
+    fn test_if_expression_rule() {
+        let circuit = "
+        machine fibo(signal n) (signal b: field) {
+            // n and be are created automatically as shared
+            // signals
+            signal a: field, i;
+
+            // there is always a state called initial
+            // input signals get bound to the signal
+            // in the initial state (first instance)
+            state initial {
+             signal c;
+
+             i, a, b, c <== 1, 1, 1, 2; // wrong
+
+             if false {
+                a <== 1;
+             }
+
+             -> middle {
+              a', b', n' <== b, c, n;
+             }
+            }
+
+            state middle {
+             signal c;
+             signal t: bool;
+
+             c <== a + b;
+
+             t <== true;
+
+             if i + 1 {                 // wrong
+                if c {                  // wrong
+                    a <== 1;
+                }
+                if t {                  // allowed
+                    a <== 1;
+                }
+                if 4 {                  // wrong
+                    a <== 1;
+                }
+              -> final {
+               i', b', n' <== i + 1, c, n;
+              }
+             } else {
+              -> middle {
+               i', a', b', n' <== i + 1, b, c, n;
+              }
+             }
+            }
+
+            // There is always a state called final.
+            // Output signals get automatically bound to the signals
+            // with the same name in the final step (last instance).
+            // This state can be implicit if there are no constraints in it.
+           }
+        ";
+
```

Comment:
You should also check that it passes with things like `3+i == a`; because that is a valid logical expression, I would just look at the top operator to see if it is logical or go with something more advance that recursively checks. But it is more complicated than currently. `==` root is a logical expression, but sides should be airthmetic expressions

Reply:
Oki I got it now! I was thinking that we could not use any field signal inside of the condition, that's why I was checking recursively. But what I understood now is that the rule just needs to make sure the condition should be a logic expr, true, false, or bool signal. Changed now for just checking the root expression of the condition